Generate serialization for SecKeychainItemRef

mscottapple opened the pull request:
#### edf8486e0ced3bf3f416bd255aa8b694d7ba01c5
<pre>
Generate serialization for SecKeychainItemRef
<a href="https://bugs.webkit.org/show_bug.cgi?id=268182">https://bugs.webkit.org/show_bug.cgi?id=268182</a>
<a href="https://rdar.apple.com/121676431">rdar://121676431</a>

Reviewed by achristensen07 (Alex Christensen).

* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/Shared/cf/ArgumentCodersCF.cpp:
(IPC::ArgumentCoder&lt;SecKeychainItemRef&gt;::encode): Deleted.
(IPC::ArgumentCoder&lt;RetainPtr&lt;SecKeychainItemRef&gt;&gt;::decode): Deleted.
* Source/WebKit/Shared/cf/ArgumentCodersCF.h:
* Source/WebKit/Shared/cf/CFTypes.serialization.in:
* Source/WebKit/Shared/cf/CoreIPCSecKeychainItem.h: Added.
(WebKit::CoreIPCSecKeychainItem::CoreIPCSecKeychainItem):
(WebKit::CoreIPCSecKeychainItem::createSecKeychainItem const):
(WebKit::CoreIPCSecKeychainItem::dataReference const):
(WebKit::CoreIPCSecKeychainItem::persistentRefForKeychainItem const):
* Source/WebKit/Shared/cf/CoreIPCSecKeychainItem.serialization.in: Added.
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/IPC/IPCSerialization.mm:
(getTempKeychain):
(destroyTempKeychain):
(TEST):

Canonical link: <a href="https://commits.webkit.org/274043@main">https://commits.webkit.org/274043@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2bc7b6145342ab109e1ce8b1dbe24096dddd98fb

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/37631 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/16509 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/39877 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/40160 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/33499 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/38948 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/19243 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/13670 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/31878 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/38196 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/13956 "Found 1 new test failure: media/video-ended-does-not-hold-sleep-assertion.html (failure)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/32975 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/12132 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/12126 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/33657 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/41421 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/34078 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/34081 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/37973 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/12662 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/10196 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/36142 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/14098 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/13062 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/4887 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/13411 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

mscottapple commented on Source/WebKit/Shared/cf/CoreIPCSecKeychainItem.h:
```diff
@@ -0,0 +1,99 @@
+
+    RetainPtr<SecKeychainItemRef> createSecKeychainItem() const
+    {
+        RELEASE_ASSERT(hasProcessPrivilege(ProcessPrivilege::CanAccessCredentials));
```
NB: This was a `RELEASE_ASSERT` in the original code.

mscottapple commented on Source/WebKit/Shared/cf/CoreIPCSecKeychainItem.h:
```diff
@@ -0,0 +1,99 @@
+        // SecKeychainItemCopyFromPersistentReference() cannot handle 0-length CFDataRefs.
+        if (CFDataGetLength(data)) {
+            ALLOW_DEPRECATED_DECLARATIONS_BEGIN
+            SecKeychainItemCopyFromPersistentReference(data, &keychainItem);
```
There aren't any API/SPI to manually deconstruct a `SecKeychainItem` into component parts. Fortunately, a persistent reference is like an alias or PK to the original item—not the item itself—so there _should_ be less room for mischief with attacker data.

achristensen07 commented on Source/WebKit/WebKit.xcodeproj/project.pbxproj:
```diff
@@ -3,7 +3,7 @@
 	classes = {
 	};
-	objectVersion = 54;
+	objectVersion = 55;
```
This change shouldn't be included in this PR.

achristensen07 commented on Tools/TestWebKitAPI/Tests/IPC/IPCSerialization.mm:
```diff
@@ -412,6 +415,35 @@ String certDataBase64(""
+    char pass[PATH_MAX];
+
+    uuid_generate_random(uu);
+    uuid_unparse_upper(uu, pass);
```
pass's length of PATH_MAX is a little strange.  Why not just use NSUUID.UUID and %@?

mscottapple replied:
Because I also use `pass` a few lines later as the C-string password for the keychain. (This is code from a previous life which I basically copied/pasted.) I could use a value smaller than `PATH_MAX` to be sure, but that's been the historical value to use when doing anything path-related.)

achristensen07 commented on Tools/TestWebKitAPI/Tests/IPC/IPCSerialization.mm:
```diff
@@ -412,6 +415,35 @@ String certDataBase64(""
+    [NSFileManager.defaultManager removeItemAtPath:path error:NULL];
+    RELEASE_ASSERT(SecKeychainCreate(path.fileSystemRepresentation, passLen, pass, false, NULL, &keychainRef) == errSecSuccess);
+    RELEASE_ASSERT(keychainRef);
+    RELEASE_ASSERT(SecKeychainUnlock(keychainRef, passLen, pass, TRUE) == errSecSuccess);
```
Do we need a corresponding SecKeychainLock call?  Also, RELEASE_ASSERT is probably a bit excessive here.  EXPECT_TRUE, EXPECT_NOT_NULL, or EXPECT_EQ cause a test failure without a crash.

mscottapple replied:
There's no need to lock this keychain, it gets destroyed instead. This is just a throw-away keychain for API which needs a keychain to operate correctly.

I will change the macros.

achristensen07 commented on Source/WebKit/Shared/cf/CoreIPCSecKeychainItem.h:
```diff
@@ -0,0 +1,99 @@
+
+    IPC::DataReference dataReference() const
+    {
+        RELEASE_ASSERT(m_persistentRef);
```
We probably don't want to introduce a crash here.  Just returning an empty data reference is a reasonable behavior here.  The hasProcessPrivilege assertions are ok though.

achristensen07 commented on Source/WebKit/Shared/cf/CoreIPCSecKeychainItem.serialization.in:
```diff
@@ -0,0 +1,31 @@
+# Copyright (C) 2023 Apple Inc. All rights reserved.
```
2024

achristensen07 commented on Tools/TestWebKitAPI/Tests/IPC/IPCSerialization.mm:
```diff
@@ -644,6 +676,29 @@ String certDataBase64(""
+    RELEASE_ASSERT(SecItemExport(cert.get(), kSecFormatX509Cert, kSecItemPemArmour, nil, &certData) == errSecSuccess);
+    CFArrayRef items = NULL;
+    RELEASE_ASSERT(SecKeychainItemImport(certData, CFSTR(".pem"), NULL, NULL, 0, NULL, tempKeychain, &items) == errSecSuccess);
+    RELEASE_ASSERT(items);
```
EXPECT_NOT_NULL.

achristensen07 commented on Source/WebKit/Shared/cf/CoreIPCSecKeychainItem.h:
```diff
@@ -0,0 +1,104 @@
+        SecKeychainItemRef keychainItem = NULL;
+        SecKeychainItemCopyFromPersistentReference(data, &keychainItem);
+        ALLOW_DEPRECATED_DECLARATIONS_END
+        return (keychainItem ? adoptCF(keychainItem) : nullptr);
```
This is an unnecessary null check.  adoptCF can be called on null.

mscottapple committed:
Generate serialization for SecKeychainItemRef
https://bugs.webkit.org/show_bug.cgi?id=268182
rdar://121676431

Reviewed by achristensen07 (Alex Christensen).

* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/Shared/cf/ArgumentCodersCF.cpp:
(IPC::ArgumentCoder<SecKeychainItemRef>::encode): Deleted.
(IPC::ArgumentCoder<RetainPtr<SecKeychainItemRef>>::decode): Deleted.
* Source/WebKit/Shared/cf/ArgumentCodersCF.h:
* Source/WebKit/Shared/cf/CFTypes.serialization.in:
* Source/WebKit/Shared/cf/CoreIPCSecKeychainItem.h: Added.
(WebKit::CoreIPCSecKeychainItem::CoreIPCSecKeychainItem):
(WebKit::CoreIPCSecKeychainItem::createSecKeychainItem const):
(WebKit::CoreIPCSecKeychainItem::dataReference const):
(WebKit::CoreIPCSecKeychainItem::persistentRefForKeychainItem const):
* Source/WebKit/Shared/cf/CoreIPCSecKeychainItem.serialization.in: Added.
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/IPC/IPCSerialization.mm:
(getTempKeychain):
(destroyTempKeychain):
(TEST):

Canonical link: https://commits.webkit.org/274043@main
